### PR TITLE
Generic params as arguments

### PR DESCRIPTION
**Pull request by petervdonovan**
Allow parameters of generic reactors to syntactically appear as arguments of contained reactors.

### CODE REVIEW

File: org.lflang/src/org/lflang/generator/c/TypeParameterizedReactor.java [hunk 1]
```diff
@@ -76,9 +90,23 @@ public InferredType resolveType(InferredType t) {
     return InferredType.fromAST(resolveType(t.astType));
   }
 
+  public String uniqueName() {
+    String name = reactor.getName().toLowerCase();
+    if (uniqueNames.containsKey(this)) return uniqueNames.get(this);
+    if (nameCounts.containsKey(name)) {
+      int currentCount = nameCounts.get(name);
+      nameCounts.put(name, currentCount + 1);
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/c/TypeParameterizedReactor.java:98]: I noticed that there is one shared counter for all names. Not that it matters, but would it be worth it to have a counter per name for the same of readability?

File: org.lflang/src/org/lflang/generator/c/TypeParameterizedReactor.java [hunk 2]
```diff
@@ -76,9 +90,23 @@ public InferredType resolveType(InferredType t) {
     return InferredType.fromAST(resolveType(t.astType));
   }
 
+  public String uniqueName() {
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/c/TypeParameterizedReactor.java:93]: Comment missing.

File: org.lflang/src/org/lflang/generator/c/CUtil.java [hunk 3]
```diff
@@ -135,7 +137,7 @@ public static String channelIndexName(PortInstance port) {
    * (to allow for instantiations that have the same name as the main reactor or the .lf file).
    */
   public static String getName(TypeParameterizedReactor reactor) {
-    String name = reactor.reactor().getName().toLowerCase() + reactor.hashCode();
+    String name = reactor.uniqueName();
```
Comment (lhstrh) [inline org.lflang/src/org/lflang/generator/c/CUtil.java:140]: This is a nice improvement!

**Review by lhstrh — COMMENTED** [review]
Looks good! Only nitpicks...

File: org.lflang/src/org/lflang/generator/c/TypeParameterizedReactor.java (hunk 1 above)
Comment (petervdonovan) [reply org.lflang/src/org/lflang/generator/c/TypeParameterizedReactor.java:102]: But I think I do use different counters for different names. `currentCount` is retrieved from a hash table the keys of which are the names.

File: org.lflang/src/org/lflang/generator/c/TypeParameterizedReactor.java (hunk 2 above)
Comment (petervdonovan) [reply org.lflang/src/org/lflang/generator/c/TypeParameterizedReactor.java:97]: Done!

File: org.lflang/src/org/lflang/generator/c/TypeParameterizedReactor.java (hunk 1 above)
Comment (lhstrh) [reply org.lflang/src/org/lflang/generator/c/TypeParameterizedReactor.java:102]: Oh, you're right. I missed that. Great!

**Review by lhstrh — APPROVED** [review]

**Pull request merged**